Feature/building names

Raycoms opened the pull request:
Closes #3130

# Changes proposed in this pull request:
- Allow to name certain huts individually (use cases will come later)

Review please

Raycoms committed:
First setup for edit name

Raycoms committed:
Finish building custom naming

Nightenom commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutNameEntry.java:
```diff
@@ -0,0 +1,62 @@
+public class WindowHutNameEntry extends Window implements ButtonHandler
+{
+    private static final String BUTTON_DONE                   = "done";
+    private static final String BUTTON_CANCEL                 = "cancel";
```
aren't these in window constants?

Raycoms replied:
Will check

Nightenom commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutNameEntry.java:
```diff
@@ -0,0 +1,62 @@
+            if (!name.isEmpty())
+            {
+                building.setCustomName(name);
+            }
```
we should add an error message if checking fails - I would just setText("Empty names are not allowed!");

Raycoms replied:
Good idea

Nightenom replied:
Why the empty check disappeared?

Nightenom commented on src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java:
```diff
@@ -0,0 +1,110 @@
+    /**
+     * The max length of the name.
+     */
+    private static final int MAX_NAME_LENGTH = 25;
```
this should be also errored when checking correctness of the input name in the window

Raycoms replied:
Here I agree

Nightenom commented on src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java:
```diff
@@ -0,0 +1,110 @@
+    {
+        super();
+        this.colonyId = colony.getID();
+        this.name = (name.length() <= MAX_NAME_LENGTH) ? name : name.substring(0, SUBSTRING_LENGTH);
```
isn't substring end exclusive?

> The substring begins at the specified beginIndex and extends to the character at index endIndex - 1.

Raycoms replied:
Good point

Nightenom commented on src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java:
```diff
@@ -0,0 +1,110 @@
+
+            if (b != null)
+            {
+                message.name = (message.name.length() <= MAX_NAME_LENGTH) ? message.name : message.name.substring(0, SUBSTRING_LENGTH);
```
why do you check the length twice?

Raycoms replied:
ugh, copy and paste error

Nightenom commented on src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java:
```diff
@@ -0,0 +1,110 @@
+            {
+                message.name = (message.name.length() <= MAX_NAME_LENGTH) ? message.name : message.name.substring(0, SUBSTRING_LENGTH);
+                b.setCustomBuildingName(message.name);
+                MineColonies.getNetwork().sendToAll(message);
```
this message does not seem to have a client acceptor method to be handled by messageOnClientThread

Raycoms replied:
ugh, copy and paste error

Raycoms replied:
Uhm, it should only be handled on server side and not on client side

Nightenom replied:
This one

Raycoms committed:
adapt to review

Nightenom commented on src/main/resources/assets/minecolonies/lang/en_us.lang:
```diff
@@ -917,4 +917,8 @@ com.minecolonies.coremod.gui.workerHuts.Sifter=Sifter
-com.minecolonies.coremod.sifter.toomuch=Set to the current daily max of %d.
\ No newline at end of file
+com.minecolonies.coremod.sifter.toomuch=Set to the current daily max of %d.
+com.minecolonies.coremod.gui.building.rename.title=Rename Your Building
```
Please include non-empty and the max length there

Nightenom reviewed (changes requested):
The max length thing - just cap the input at 15 chars and mention it inside the `com.minecolonies.coremod.gui.building.rename.title`
We may allow longer ones with new assets (I checked that 15 is max now)

Raycoms committed:
remove sendToAll

Raycoms commented:
Removed send to all, I removed us empty because empty is valid, empty
resets to default

On Tue, Mar 12, 2019, 11:30 Nightenom <notifications@github.com> wrote:

> *@Nightenom* requested changes on this pull request.
>
> The max length thing - just cap the input at 15 chars and mention it
> inside the com.minecolonies.coremod.gui.building.rename.title
> We may allow longer ones with new assets (I checked that 15 is max now)
> ------------------------------
>
> In
> src/main/java/com/minecolonies/coremod/client/gui/WindowHutNameEntry.java
> <https://github.com/ldtteam/minecolonies/pull/3489#discussion_r264703998>:
>
> > +    @Override
> +    public void onOpened()
> +    {
> +        findPaneOfTypeByID(INPUT_NAME, TextField.class).setText(building.getCustomName());
> +    }
> +
> +    @Override
> +    public void onButtonClicked(@NotNull final Button button)
> +    {
> +        if (button.getID().equals(BUTTON_DONE))
> +        {
> +            final String name = findPaneOfTypeByID(INPUT_NAME, TextField.class).getText();
> +            if (!name.isEmpty())
> +            {
> +                building.setCustomName(name);
> +            }
>
> Why the empty check disappeared?
> ------------------------------
>
> In
> src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java
> <https://github.com/ldtteam/minecolonies/pull/3489#discussion_r264705553>:
>
> > +        BlockPosUtil.writeToByteBuf(buf, buildingId);
> +    }
> +
> +    @Override
> +    public void messageOnServerThread(final HutRenameMessage message, final EntityPlayerMP player)
> +    {
> +        final Colony colony = ColonyManager.getColonyByDimension(message.colonyId, message.dimension);
> +        if (colony != null && colony.getPermissions().hasPermission(player, Action.MANAGE_HUTS))
> +        {
> +            final AbstractBuilding b = colony.getBuildingManager().getBuildings().get(message.buildingId);
> +
> +            if (b != null)
> +            {
> +                message.name = (message.name.length() <= MAX_NAME_LENGTH) ? message.name : message.name.substring(0, SUBSTRING_LENGTH);
> +                b.setCustomBuildingName(message.name);
> +                MineColonies.getNetwork().sendToAll(message);
>
> This one
> ------------------------------
>
> In src/main/resources/assets/minecolonies/lang/en_us.lang
> <https://github.com/ldtteam/minecolonies/pull/3489#discussion_r264706496>:
>
> > \ No newline at end of file
> +com.minecolonies.coremod.sifter.toomuch=Set to the current daily max of %d.
> +com.minecolonies.coremod.gui.building.rename.title=Rename Your Building
>
> Please include non-empty and the max length there
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ldtteam/minecolonies/pull/3489#pullrequestreview-213420702>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGI9y--o-TykM-JTPosjoh15TEjV3sNpks5vV7oHgaJpZM4bo3g5>
> .
>

Raycoms committed:
Merge remote-tracking branch 'origin/version/1.12' into feature/building-names
